source/_posts/2016-12-29-testing-against-external-services.markdown

```diff
@@ -0,0 +1,91 @@
+
+- The gem hasn’t been maintained in a couple years, which started to cause intermittent spec failures.
+- The gem doesn't support SFTP connections, which our service relies much more heavily on than FTP.
+- The gem doesn't actually mock anything--it starts up a local FTP server, runs the tests, and then shuts the server down. Each test that’s run against it is generating a real response, albeit locally. This not only took up a lot of time (starting and stopping the local server before and after each spec), but was also another source of intermittent failures.
+
+So, after creating a plan of attack, I rewrote or deleted every single test that used this gem. Each test was also re-evaluated for "sighting along the space capsule," as well—any test that was written for a private method, a message sent to the object’s self, or an implementation detail was scrutinized (and most of them deleted), which exposed the true underlying issue: almost every single test we'd written that used the gem was an *implementation* test, not an *interface* test.
+
+External responses are supposed to be completely mocked out in testing, not make actual calls to an API or server. Tests exist to make sure OUR code works as expected, not to verify that no errors are raised when making external requests. A test failing because an external response was incorrect is a broken test already.
+
+##### Reframing our point of view
+
+Here’s an example to illustrate my point. Let’s say we have an object whose job is to get a list of files from a specific folder on an FTP server. For brevity’s sake, let’s define it like this with the nitty gritty details fleshed out elsewhere:
+
+    class FtpThing
+      def getListOfFiles(username, password, folderName)
+        @someConnection.open(username, password)
+        listOfFiles = @someConnection.list(folderName)
+        @someConnection.close
+        listOfFiles
+      end
+    end
+
+Now, one option for testing this is the way we did it before, something like this:
+
+    let(:ftp_thing) { FtpThing.new }
+    let(:list) { [‘file1.txt’, ‘file2.txt'] }
+
+    before do
+      # spin up the “mock" (real) server
+      # put a (real) file or two in a folder on the server
+    end
+
+    it ‘gives me the right list of files’ do
+      expect(ftp_thing.getListOfFiles(‘username’, ‘password’, ‘folderName’)).to eq(list)
+      ftpThing.getListOfFiles(‘username’, ‘password’, ‘folderName’)
+    end
+
+    after do
+      # shut down the server
+    end
+
+On the surface, this doesn’t look terrible. It’s testing the value returned from `getListOfFiles`, right? Yay that’s what we want! But… not quite. We’ve got a (real) server opened, and we’re hoping that for whatever reason our local server doesn’t fail or raise any errors and that’s how our test will pass. Instead, we need to allow a mocked response and set up our expectation this way:
```

Comment:
* `Yay, that's what we want!`
* Suggestion: `We've started a real server and hope it started in time because are tests are about run` (I think active voice brings the reader in)